storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -4017,122 +4014,1889 @@ dberr_t fil_space_t::garbage_collect(bool shutdown)
+        continue;
+      m_extent_info[xdes_n]= xdes_get_state(descr);
+    }
+    goto func_exit;
+  }
+
+  /** Find the earlier free extent for the given used extent
+  @param max_limit Find the extent below max limit extent
+  @return value
+  @retval FIL_NULL if there is no extent */
+  uint32_t find_free_extent(uint32_t max_limit) noexcept
+  {
+    for (auto &extent_info : m_extent_info)
+    {
+      if (max_limit <= extent_info.first)
+        return FIL_NULL;
+
+      if (extent_info.second == XDES_FREE)
+      {
+        /* Mark the extent as used one */
+        extent_info.second = XDES_USED;
+        return extent_info.first;
+      }
+    }
+    return FIL_NULL;
+  }
+
+  /** Defragment the indexes */
+  dberr_t defragment_index(dict_index_t &index) noexcept
+  {
+    IndexDefragmenter index_defrag(index);
+    return index_defrag.defragment(this);
+  }
+
+  /** Defragment the table */
+  dberr_t defragment_table(const dict_table_t *table) noexcept
+  {
+    for (dict_index_t *index= dict_table_get_first_index(table);
+         index; index= dict_table_get_next_index(index))
+    {
+      dberr_t err= fseg_validate(fil_system.sys_space, index);
+      if (err == DB_SUCCESS)
+        err= defragment_index(*index);
+
+      if (err)
+      {
+        sql_print_error("InnoDB: Defragmentation of %s in %s failed: %s",
+                        index->name, table->name.m_name, ut_strerr(err));
+        return err;
+      }
+    }
+    return DB_SUCCESS;
+  }
+public:
+  const uint32_t m_extent_size;
+
+  SpaceDefragmenter() noexcept : m_extent_size(FSP_EXTENT_SIZE) {}
+
+  /** Find the new extent for the existing last used extent
+  Iterate the tablespace from last and find out the free
+  extent in the beginning of the tablespace */
+  dberr_t find_new_extents() noexcept
+  {
+    dberr_t err= extract_extent_state();
+    if (err) return err;
+
+    uint32_t free_limit= fil_system.sys_space->free_limit;
+    uint32_t fixed_size= srv_sys_space.get_min_size();
+    while (free_limit > fixed_size)
+    {
+      uint32_t state= m_extent_info[free_limit];
+
+      switch (state) {
+      case XDES_USED:
+        goto func_exit;
+      case XDES_FREE:
+        goto prev_extent;
+      case XDES_FSEG:
+      case XDES_FULL_FRAG:
+      case XDES_FREE_FRAG:
+        uint32_t dest= find_free_extent(free_limit);
+        if (dest == FIL_NULL)
+          goto func_exit;
+        m_extent_map[free_limit]= dest;
+        break;
+      }
+prev_extent:
+      free_limit-= FSP_EXTENT_SIZE;
+    }
+func_exit:
+    if (m_extent_map.empty())
+      return DB_SUCCESS_LOCKED_REC;
+
+    sql_print_information("InnoDB: System tablespace defragmentation "
+                          "process starts");
+    sql_print_information("InnoDB: Moving the data from extent %"
+                          PRIu32 " to %" PRIu32, m_extent_map.begin()->first,
+                          m_extent_map.rbegin()->first);
```

Comment:
This message is better now, but the wording "moving… from extent 4096 to 22016" is misleading. Better: "from extents 4096 through 22016". We’re not saying where it is being moved to.

Can you make one of the tests search for this pattern? The message being is output in `innodb.sys_defragment` and `innodb.sys_defragment_fail`.